rules/Renaming/Rector/FunctionLike/RenameFunctionLikeParamWithinCallLikeArgRector.php

```diff
@@ -0,0 +1,281 @@
+
+        $this->renameFunctionLikeParamWithinCallLikeParams = $configuration;
+    }
+
+    private function processFunctionLike(
+        CallLike $callLike,
+        RenameFunctionLikeParamWithinCallLikeArg $renameFunctionLikeParamWithinCallLikeArg
+    ): void {
+        if (is_int($renameFunctionLikeParamWithinCallLikeArg->getCallLikePosition())) {
+            $arg = $this->processPositionalArg($callLike, $renameFunctionLikeParamWithinCallLikeArg);
+        } else {
+            $arg = $this->processNamedArg($callLike, $renameFunctionLikeParamWithinCallLikeArg);
+        }
+
+        if (! $arg instanceof Arg) {
+            return;
+        }
+
+        $functionLike = $arg->value;
+        if (! $functionLike instanceof FunctionLike) {
+            return;
+        }
+
+        if (! isset($functionLike->params[$renameFunctionLikeParamWithinCallLikeArg->getFunctionLikePosition()])) {
+            return;
+        }
+
+        $this->refactorParameter(
+            $functionLike->params[$renameFunctionLikeParamWithinCallLikeArg->getFunctionLikePosition()],
+            $functionLike,
+            $renameFunctionLikeParamWithinCallLikeArg
+        );
+    }
+
+    /**
+     * Rename the Parameter variable name
+     */
+    private function refactorParameter(
+        Param $param,
+        FunctionLike $functionLike,
+        RenameFunctionLikeParamWithinCallLikeArg $renameFunctionLikeParamWithinCallLikeArg
+    ): void {
+        if (! $param->var instanceof Variable) {
+            return;
+        }
+
+        if (! is_string($param->var->name)) {
+            return;
+        }
+
+        $oldName = $param->var->name;
+
+        // skip if the name is in use within the context of the function like
+        if ($this->isVariableNameUsedInFunctionLike(
+            $functionLike,
+            $oldName,
+            $renameFunctionLikeParamWithinCallLikeArg->getNewParamName()
+        )) {
+            return;
+        }
+
+        $param->var->name = $renameFunctionLikeParamWithinCallLikeArg->getNewParamName();
+
+        // refactor the FunctionLike usage of the variable
+        $this->traverseNodesWithCallable($functionLike, function (Node $node) use (
+            $oldName,
+            $renameFunctionLikeParamWithinCallLikeArg
+        ): ?Node {
+            if (! $node instanceof Variable) {
+                return null;
+            }
+
+            if (! $this->isName($node, $oldName)) {
+                return null;
+            }
+
+            $node->name = $renameFunctionLikeParamWithinCallLikeArg->getNewParamName();
+
+            return $node;
+        });
+
+        $this->hasChanged = true;
+    }
+
+    private function isVariableNameUsedInFunctionLike(
+        FunctionLike $functionLike,
+        string $oldName,
+        string $newName
+    ): bool {
+        $isUsed = false;
+        $this->traverseNodesWithCallable($functionLike, function (Node $node) use (
```

Comment:
you actually should traverse its stmts instead of re-iterate it:

```suggestion
        $this->traverseNodesWithCallable((array) $functionLike->getStmts(), function (Node $node) use (
```

so `$node !== $functionLike` should never needed.

Reply:
@samsonasik I played around with your suggestion and found it to be a problem because it then doesn't pick up scenarios such as, is there already a parameter in the function with the same name or a parameter "used" by a Closure within the FunctionLike. When adding these to a larger array it seemed to break further. This really seems to be the simplest option.

```diff
-SomeClass::someCall(function ($query) {
-    echo $query;
+SomeClass::someCall(function ($qry) {
+    echo $qry;
     return function ($query) {
         return $query;
     };
```

```diff
-SomeClass::someCall(function ($qry) use ($query) {
-    return $qry;
+SomeClass::someCall(function ($query) use ($query) {
+    return $query;
 });
```

As you can imagine both of these operations would lead to broken PHP.

Reply:
Ok 👍